### ISSUE TITLE
yarn upgrade not rewriting yarn.lock file when using yarn workspaces

### ISSUE THREAD

**Issue by josh08h** [yarnpkg/yarn#7698]
Full reproduction found here https://github.com/josh08h/yarn_workspaces_upgrade_example

**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
Running `yarn upgrade webpack` or `yarn upgrade webpack --force` is not upgrading `webpack` even though the semver range specified in the `package.json` file allows for a new version.

**If the current behavior is a bug, please provide the steps to reproduce.**
```
// package.json

"dependencies": {
      "webpack": "~4"
    }
```

```
// yarn.lock
webpack@~4:
  version "4.29.3"
  resolved "https://registry.yarnpkg.com/webpack/-/webpack-4.29.3.tgz#e0b406a7b4201ed5e4fb4f84fd7359f9a7db4647"
  integrity sha512-xPJvFeB+8tUflXFq+OgdpiSnsCD5EANyv56co5q8q8+YtEasn5Sj3kzY44mta+csCIEB0vneSxnuaHkOL2h94A==
  dependencies:
............
............
............

```

**What is the expected behavior?**
I would expect `yarn upgrade webpack` to pull in the latest that my `package.json` allows and rewrite the `yarn.lock` file.

```
Package                           Current Wanted  Latest
webpack                           4.29.3  4.41.2  4.41.2
```

But instead I see:
```
[5/5] 🔨  Rebuilding all packages...
success Saved lockfile.
success Saved 0 new dependencies.
```

**Please mention your node.js, yarn and operating system version.**
```
yarn --version
1.19.1
node --version
v10.15.3
```

**Comment by josh08h** [comment]
Note I am using yarn workspaces.

**Comment by patrick-mcdougle** [comment]
I have attempted to reproduce this issue and cannot. Please note that I am not a maintainer of yarn, just trying to be helpful to a maintainer and/or the original poster (OP).

Here's what I did to try...in a new folder running the same version of yarn and node v10.16.3

Setup:
```
yarn init (all defaults)
yarn add webpack@4.29.3 (just to set things up)
manually edit package.json to change webpack 4.29.3 to ~4 also edit yarn.lock to change from webpack@4.29.3: to webpack@~4:
yarn install (just to get the lockfile up to date)
```

At this point yarn seems to be in the state described by OP. 
`yarn outdated` shows the same output as OP.
yarn.lock shows the same version, integrity and resolved all match OP's details.

Running `yarn upgrade webpack` adds 244 new dependencies and is updated to latest in the yarn.lock file.

I do not know what yarn workspaces is, so I probably am not using them in this attempted repro.

Sharing your whole package.json + yarn.lock file might be in order...

**Comment by josh08h** [comment]
Hi @patrick-mcdougle 

Thanks for investigating. I've created an example repo which exhibits the behaviour I've described above.

https://github.com/josh08h/yarn_workspaces_upgrade_example

**Comment by josh08h** [comment]
Thanks for taking a look into this @patrick-mcdougle..

There is a subtle difference in the approach that you have taken to reproduce this example. Your `yarn.lock` file will have:

```
webpack@4.29.3:
  version "4.29.3"
  resolved "https://registry.yarnpkg.com/webpack/-/webpack-4.29.3.tgz#e0b406a7b4201ed5e4fb4f84fd7359f9a7db4647"
```
as the entry for Webpack, whereas mine has:

```
webpack@~4:
  version "4.29.3"
  resolved "https://registry.yarnpkg.com/webpack/-/webpack-4.29.3.tgz#e0b406a7b4201ed5e4fb4f84fd7359f9a7db4647"
```
notice we have both resolved to the same `4.29.3`, however mine has `webpack@~4` and yours does not since when I ran `yarn install` last I was specifying `"webpack": "~4"` in my `package.json` and so it resolved to `4.29.3` which was the highest `wanted` at the time.

I've produced an example repo here demoing the issue https://github.com/josh08h/yarn_workspaces_upgrade_example

**Comment by patrick-mcdougle** [comment]
@josh08h I thought I covered this case in my 3rd and 4th steps where I manually edited both package.json and yarn.lock (and ran yarn install to get it to stop complaining about the lockfile being outdated).
> manually edit package.json to change webpack 4.29.3 to ~4 also edit yarn.lock to change from webpack@4.29.3: to webpack@~4:

After doing so I verified that yarn.lock was in the same state as your lock file.

I guess what my reproduction attempt shows is that WITHOUT using workspaces, the bug does not present itself, so the bug is somewhere in the workspaces feature.
